pkg/kubectl/cmd/util/openapi/validation/types.go

```diff
@@ -0,0 +1,265 @@
+}
+
+// AddValidationError wraps the given error into a ValidationError and
+// attaches it to this item.
+func (b *baseItem) AddValidationError(err error) {
+	b.errors.AppendErrors(ValidationError{Path: b.path.String(), Err: err})
+}
+
+// AddError adds a regular (non-validation related) error to the list.
+func (b *baseItem) AddError(err error) {
+	b.errors.AppendErrors(err)
+}
+
+// CopyErrors adds a list of errors to this item. This is useful to copy
+// errors from subitems.
+func (b *baseItem) CopyErrors(errs []error) {
+	b.errors.AppendErrors(errs...)
+}
+
+// Path returns the path of this item, helps print useful errors.
+func (b *baseItem) Path() *openapi.Path {
+	return &b.path
+}
+
+// mapItem represents a map entry in the yaml.
+type mapItem struct {
+	baseItem
+
+	Map map[string]interface{}
+}
+
+func (mi *mapItem) sortedKeys() []string {
+	sortedKeys := []string{}
+	for key := range mi.Map {
+		sortedKeys = append(sortedKeys, key)
+	}
+	sort.Strings(sortedKeys)
+	return sortedKeys
+}
+
+var _ ValidationItem = &mapItem{}
+
+func (mi *mapItem) VisitPrimitive(p *openapi.Primitive) {
+	mi.AddValidationError(InvalidTypeError{Path: p.GetPath().String(), Expected: p.Type, Actual: "map"})
+}
+
+func (mi *mapItem) VisitArray(a *openapi.Array) {
+	mi.AddValidationError(InvalidTypeError{Path: a.GetPath().String(), Expected: "array", Actual: "map"})
+}
+
+func (mi *mapItem) VisitMap(m *openapi.Map) {
+
+	for _, key := range mi.sortedKeys() {
+		item, err := itemFactory(mi.Path().FieldPath(key), mi.Map[key])
+		if err != nil {
+			mi.AddError(err)
+			continue
+		}
+		m.SubType.Accept(item)
+		mi.CopyErrors(item.Errors())
+	}
+}
+
+func (mi *mapItem) VisitKind(k *openapi.Kind) {
+	// Verify each sub-field.
+	for _, key := range mi.sortedKeys() {
+		item, err := itemFactory(mi.Path().FieldPath(key), mi.Map[key])
+		if err != nil {
+			mi.AddError(err)
+			continue
+		}
+		if _, ok := k.Fields[key]; !ok {
+			mi.AddValidationError(UnknownFieldError{Path: k.GetPath().String(), Field: key})
+			continue
+		}
+		k.Fields[key].Accept(item)
+		mi.CopyErrors(item.Errors())
+	}
+
+	// Verify that all required fields are present.
+	for _, required := range k.RequiredFields {
+		if _, ok := mi.Map[required]; !ok {
+			mi.AddValidationError(MissingRequiredFieldError{Path: k.GetPath().String(), Field: required})
+		}
+	}
+}
+
+// arrayItem represents a yaml array.
+type arrayItem struct {
+	baseItem
+
+	Array []interface{}
+}
+
+var _ ValidationItem = &arrayItem{}
+
+func (ai *arrayItem) VisitPrimitive(p *openapi.Primitive) {
+	ai.AddValidationError(InvalidTypeError{Path: p.GetPath().String(), Expected: p.Type, Actual: "array"})
+}
+
+func (ai *arrayItem) VisitArray(a *openapi.Array) {
+	for i, v := range ai.Array {
+		item, err := itemFactory(ai.Path().ArrayPath(i), v)
+		if err != nil {
+			ai.AddError(err)
+			continue
+		}
+		a.SubType.Accept(item)
+		ai.CopyErrors(item.Errors())
+	}
+}
+
+func (ai *arrayItem) VisitMap(m *openapi.Map) {
```

Comment:
nit: you could probably reduce code duplication by creation a base impl that takes a type and emits and error for everything, then embed it and override.